src/Writer/Factory/WriterFactory.php

```diff
@@ -0,0 +1,126 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace Zend\Log\Writer\Factory;
+
+use Interop\Container\ContainerInterface;
+use Zend\ServiceManager\Exception\InvalidServiceException;
+use Zend\ServiceManager\FactoryInterface;
+use Zend\ServiceManager\ServiceLocatorInterface;
+
+/**
+ * Factory for instantiating classes with no dependencies or which accept a single array.
+ *
+ * The WriterFactory can be used for any class that:
+ *
+ * - has no constructor arguments;
+ * - accepts a single array of arguments via the constructor.
+ *
+ * It replaces the "invokables" and "invokable class" functionality of the v2
+ * service manager, and can also be used in v2 code for forwards compatibility
+ * with v3.
+ */
+final class WriterFactory implements FactoryInterface
+{
+    /**
+     * Options to pass to the constructor (when used in v2), if any.
+     *
+     * @param null|array
+     */
+    private $creationOptions;
+
+    /**
+     * @param null|array|Traversable $creationOptions
+     * @throws InvalidServiceException if $creationOptions cannot be coerced to
+     *     an array.
+     */
+    public function __construct($creationOptions = null)
+    {
+        if (null === $creationOptions) {
+            return;
+        }
+
+        $this->setCreationOptions($creationOptions);
+    }
+
+    /**
+     * Create an instance of the requested class name.
+     *
+     * @param ContainerInterface $container
+     * @param string $requestedName
+     * @param null|array $options
+     * @return object
+     */
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
+    {
+        if (isset($options['filter_manager']) && is_string($options['filter_manager'])) {
+            $options['filter_manager'] = $container->get($options['filter_manager']);
+        } elseif (!isset($options['filter_manager']) && $container->has('LogFilterManager')) {
```

Comment:
We likely want to grab the canonical `LogFilterManager` service (and the `LogFormatterManager` service, from below) if `$options['filter_manager']` is set but not a string and/or the named service does not exist. I'd move that logic into a separate method, something like this:

```php
private function getFilterManager(ContainerInterface $container, array $options)
{
    if (! isset($options['filter_manager']) && ! $container->has('LogFilterManager')) {
        return null;
    }
    
    if (! isset($options['filter_manager']) && $container->has('LogFilterManager')) {
	return $container->get('LogFilterManager');
    }
    
    if (isset($options['filter_manager'])
        && is_string($options['filter_manager'])
        && $container->has($options['filter_manager'])
    ) {
        return $container->get($options['filter_manager']);
    }
    
    throw new ServiceNotCreatedException('Invalid zend-log filter_manager option provided');
}
```    

which you would then call from `__invoke()` using:

```php
$options = (array) $options;
if (null !== ($filterManager = $this->getFilterManager($container, $options))) {
    $options['filter_manager'] = $filterManager;
}
```

You would then do the same with the formatter manager. This approach ensures that we have valid arguments, and allows us to raise a very specific exception about why we cannot create the *requested* service (invalid filter/formatter configuration).